src/lib.mo

```diff
@@ -0,0 +1,327 @@
+          case (_) {};
+        };
+      };
+    };
+
+  };
+
+  class PullValue(prefix_ : Text, pull : () -> Nat) {
+    public let prefix = prefix_;
+
+    public func dump() : [(Text, Nat)] = [(prefix # "{}", pull())];
+
+    public func share() : ?StableDataItem = null;
+    public func unshare(data : StableDataItem) = ();
+  };
+
+  class CounterValue(prefix_ : Text, isStable : Bool) {
+    public let prefix = prefix_;
+
+    public var value = 0;
+
+    public func add(n : Nat) { value += n };
+    public func set(n : Nat) { value := n };
+
+    public func dump() : [(Text, Nat)] = [(prefix # "{}", value)];
+
+    public func share() : ?StableDataItem {
+      if (not isStable) return null;
+      ? #counter(value);
+    };
+    public func unshare(data : StableDataItem) = switch (data, isStable) {
+      case (#counter x, true) value := x;
+      case (_) {};
+    };
+  };
+
+  class GaugeValue(prefix_ : Text, buckets_ : ?[Nat], now : () -> Int, watermarkResetIntervalSeconds : Nat, isStable : Bool) {
+    public let prefix = prefix_;
+    public let buckets : ?[Nat] = switch (buckets_) {
+      case (null) null;
+      case (?b) ?Array.sort(b, Nat.compare);
+    };
+
+    class WatermarkTracker<T>(default : T, condition : (old : T, new : T) -> Bool, resetIntervalSeconds : Nat) {
+      var lastWatermarkTimestamp : Time.Time = 0;
+      public var value : T = default;
+      public func update(current : T) {
+        if (condition(value, current) or now() > lastWatermarkTimestamp + resetIntervalSeconds * 1_000_000_000) {
+          value := current;
+          lastWatermarkTimestamp := now();
+        };
+      };
+    };
+
+    public var count : Nat = 0;
+    public var sum : Nat = 0;
+    public var highWatermark : WatermarkTracker<Nat> = WatermarkTracker<Nat>(0, func(old, new) = new > old, watermarkResetIntervalSeconds);
+    public var lowWatermark : WatermarkTracker<Nat> = WatermarkTracker<Nat>(0, func(old, new) = new < old, watermarkResetIntervalSeconds);
+    public var lastValue : Nat = 0;
+    public let bucketValues : [var Nat] = Array.tabulateVar<Nat>(
+      switch (buckets) { case (?b) { b.size() + 1 }; case (null) { 0 } },
+      func(n) = 0,
+    );
+
+    public func update(current : Nat) {
+      count += 1;
+      sum += current;
+      highWatermark.update(current);
+      lowWatermark.update(current);
+      switch (buckets) {
+        case (null) {};
+        case (?b) {
+          var skip : Nat = 0;
+          label l for (bucketValue in b.vals()) {
+            if (current <= bucketValue) {
+              break l;
+            } else {
+              skip += 1;
+            };
+          };
+          for (i in Iter.range(skip, bucketValues.size() - 1)) {
+            bucketValues[i] += 1;
+          };
+        };
+      };
+    };
+
+    public func dump() : [(Text, Nat)] = Array.tabulate<(Text, Nat)>(
+      4 + bucketValues.size(),
+      func(n) = switch (n) {
+        case (0)(prefix # "_sum{}", sum);
```

Comment:
Wondering if it's worth defining a small function that produces the Text part with two arguments: suffix (e.g. "sum") and an array of labels. The prefix is hard-coded in the function.

Reply:
As discussed in private conversation, postponed